### PR TITLE
Db export

### PR DESCRIPTION
**Pull request by chiahaoliu**
In progress of ``db.export`` functionality

### CODE REVIEW

File: databroker/broker.py [hunk 1]
```diff
@@ -509,10 +509,18 @@ def export(headers, mds, fs):
...
             mds : metadatasote
-                metadata store object #FIXME
+                metadata store object
             fs : file store
-                file store object #FIXME
+                file store object
         """
+        for header in headers:
+            mds.insert_run_start(**header['start'])
+            for descripto in header['descriptor']:
```
Comment (arkilic) [inline databroker/broker.py]: typo `descripto`

File: databroker/broker.py [hunk 2]
```diff
@@ -501,6 +501,28 @@ def process(self, headers, func, fields=None, fill=False):
...
+            fs : file store
+                file store object
+        """
+        for header in headers:
+            mds.insert_run_start(**header['start'])
+            for descripto in header['descriptor']:
+                mds.insert_descriptor(**descriptor)
+                events = self.get_events(header)
+                for event in events:
+                    mds.insert_event(event, descriptor)
```
Comment (danielballan) [inline databroker/broker.py]: The `insert_event` API works the same as the other insert methods: it takes `**event`. The `descriptor` arg in its signature refers to `event['descriptor']`, which is the uid for `descriptor`.

Also, this for loop should be un-indented. `get_events` returns events for *all* descriptors, in bulk . It does not need to be repeated for each descriptor.

File: databroker/broker.py (hunk 1 above)
Comment (chiahaoliu) [inline databroker/broker.py]: caught me

**Comment by chiahaoliu** [comment]
update: code was modified based on instruction and typo was fixed

**Comment by danielballan** [comment]
Looks good. Next up is FileStore. Steps will be:

1. Use the databroker method `get_resource_uids` to identify which resource documents we need to copy.
2. Get these resource documents from `self.fs` and insert them into `fs`.
3. Use a (yet to be written) method on filestore to identify which datum documents we need to copy, and insert them into `fs` as well.

At this point, the `mds` and `fs` will work correctly, but `fs` still refers to files (like TIFFs) that are on the local filesystem. We need to copy the underlying files to a new location and then update `fs` to point there instead.

4. Add an additional required parameter to the signature of `db.export` named `new_root`.
5. Copy the files into the `new_root` directory using a (yet to be written) method on filestore handlers.
6. Update the copied resource documents in the new `fs` using `change_root`.

File: databroker/broker.py [hunk 3]
```diff
@@ -565,6 +565,26 @@ def process(self, headers, func, fields=None, fill=False):
...
+    def export(headers, mds, fs):
+        """ export a list of headers
+
+            Parameters:
+            -----------
+            headers : databroker.header
+                list of headers that are going to be exported
+            mds : metadatasote
+                metadata store object
+            fs : file store
```
Comment (danielballan) [inline databroker/broker.py]: file store -> FileStore

File: databroker/broker.py [hunk 4]
```diff
@@ -565,6 +565,26 @@ def process(self, headers, func, fields=None, fill=False):
...
                  fields=fields, fill=fill)
 
+    def export(headers, mds, fs):
+        """ export a list of headers
+
+            Parameters:
+            -----------
+            headers : databroker.header
+                list of headers that are going to be exported
+            mds : metadatasote
```
Comment (danielballan) [inline databroker/broker.py]: metadatasote -> MDS
metadata store object -> target instance of metadatastore to export into

File: databroker/broker.py [hunk 5]
```diff
@@ -565,6 +565,26 @@ def process(self, headers, func, fields=None, fill=False):
...
+                metadata store object
+            fs : file store
+                file store object
```
Comment (danielballan) [inline databroker/broker.py]: file store object -> target instance of filestore to export into

**Comment by danielballan** [comment]
Other people are expressing interest in this feature. Do you have time to work with me to push this across the finish line next week?

**Comment by chiahaoliu** [comment]
Sure, I would love to finish it by next week.

File: databroker/broker.py [hunk 6]
```diff
@@ -577,6 +577,46 @@ def process(self, headers, func, fields=None, fill=False):
...
+                target instance of filestore to export info
+            new_root : str
+                optional. root directory of files that are going to
+                be exported
+        """
+        for header in headers:
+            # insert mds
+            mds.insert_run_start(**header['start'])
+            events = self.get_events(header)
+            for descripto in header['descriptor']:
```
Comment (danielballan) [inline databroker/broker.py]: typo

File: databroker/broker.py [hunk 7]
```diff
@@ -577,6 +577,46 @@ def process(self, headers, func, fields=None, fill=False):
 
     get_fields = staticmethod(get_fields)  # for convenience
 
+    def export(headers, mds, fs, new_root=None):
```
Comment (danielballan) [inline databroker/broker.py]: We can simplify this signature by just passing in a `Broker` instance that has `mds` and `fs` as attributes. I would suggest the signature: `export(headers, db, new_root=None`)

File: databroker/broker.py [hunk 8]
```diff
@@ -577,6 +577,46 @@ def process(self, headers, func, fields=None, fill=False):
...
+        """ export a list of headers
+
+            Parameters:
+            -----------
+            headers : databroker.header
+                list of headers that are going to be exported
+            mds : MDS
```
Comment (danielballan) [inline databroker/broker.py]: (update docstring accordingly)

File: databroker/broker.py (hunk 7 above)
Comment (danielballan) [inline databroker/broker.py]: Also, this should have `self` as the first arg.

**Comment by danielballan** [comment]
These commits were incorporated in #88.

**Pull request closed without merging**